Add EventApi

Nerixyz opened the pull request:
Pull request checklist:

- [ ] `CHANGELOG.md` was updated, if applicable

# Description

Most important thing first: Don't bother running the CI yet since it's bound to fail on formatting and only a draft for now.

![example image](https://i.imgur.com/MFdX4kt.png)

This PR implements the EventApi for Chatterino. 

**Why implement the EventApi in the first place?** Now that 7TV is established in many channels I often find myself reloading emotes. The EventApi is the best solution for this.

**But other platforms don't have such a thing.** I feel like 7TV channels are more dynamic in the sense that they add/update/remove 7TV emotes more often than BTTV/FFZ emotes.

**Why not add this implementation yet?** Api V3 is on the way and might change the way 7TV emotes are handled in Chatterino and the structure of EventApi actions so we should wait to merge this. _But_ it doesn't hurt to have a test implementation.

**TODOs:**

* [x] **Formatting**
* [x] **Figure out why the client sometimes doesn't seem to join channels.**
* [ ] **Only join channels with 7TV emotes.** I'm not sure where to hook when joining. I'd say the eventapi should return the channel when we failed to join it, so we can remove it from our internal list.
* [ ] ~~**Rethink how to update emotes.**~~ Since we only cache emotes by emote-id, and aliases are possible, we have to add some additional checks when searching. For example another channel might have added an emote with an alias but we want the original name, then both ids are the same but they don't map to the same emote-ptr. We also need to copy the map when adding/updating/removing emotes, because it's marked as `const`.
* [x] **Merge parts with the PubsubClient since the logic is very similar or even the same.**
* [ ] **Test many setups (emote cached/not cached, added/not added, add/remove alias...).**
* [ ] ~~**Check where we need to copy and where we are fine with references.**~~ Explained in the updating section. tl;dr we need to copy the emote map.

**Notes**

I copied a lot of code from the PubsubClient since the EventApi is very similar.

Nerixyz committed:
feat: add initial event-api support

Nerixyz committed:
fix: add types to std::min/max

I don't know why it compiled before, but now it doesn't?!

Nerixyz committed:
fix: use the correct emote name and action author

Nerixyz committed:
refactor: move join and part

Nerixyz committed:
feat: check structure of event and emotes

Nerixyz committed:
fix: use std::any_of

Nerixyz committed:
fix: qmake

hopefully

Nerixyz committed:
fix: include QJsonParseError

Nerixyz committed:
fix: formatting

AnatoleAM commented:
Good start 👍 

I'll note that the Event API is separate from the main API, and will continue to function the same after v3 deploys. Events will get a new version though as Emote Sets have fundamentally changed the concept of "channel emotes" that the event api is based on. The plan is to make event sessions mutable even via SSE (through a REST endpoint) and allow a wide range of topics clients can subscribe to.

Nerixyz committed:
refactor: use the architecture of the refactored pubsub client

This should result in the client being more stable, having a backoff when reconnecting and checking for pings.
Since SevenTv/chatterino7 hasn't updated its fork, magic_enum isn't a dependency yet.

Nerixyz committed:
fix: formatting

Nerixyz committed:
fix: add a default value for `lastPing_`

Nerixyz committed:
fix: turn actor into a link

Nerixyz committed:
fix: duplicate name `Listener`

Nerixyz committed:
chore: use a #define for the eventapi url

Nerixyz committed:
refactor: use chatterinowebsocketpplogger since it's the same

Nerixyz committed:
refactor: use magic_enum

✨

Nerixyz committed:
fix: initialize `lastPing_` in the constructor

Nerixyz committed:
chore: use longer path for include directives

Nerixyz committed:
refactor: rename Base.cpp to EventApiMessage.cpp

Hopefully this fixes the qmake error (Clueless)

Nerixyz commented:
![](https://cdn.7tv.app/emote/60aea54bf6a2c3b332395990/1x) Awesome! Green CI really made my day!

This also means you can test the PR now: [Download in the artifacts section](https://github.com/SevenTV/chatterino7/actions/runs/2314490446).

Nerixyz committed:
fix: only add unlisted emotes if settings allow it

Nerixyz committed:
feat: allow searching for add/update/remove messages

Nerixyz committed:
feat: show add/remove/delete messages in usercard

Nerixyz commented:
If the eventapi will stay the same (and the messages as well), I'll mark this PR as ready-for-review since it works for me and I can get some feedback.

<details>
<summary>This is how it currently looks like</summary>

![image](https://i.imgur.com/4Cv9bTn.png)
</details>

I'll list the current messages here, because they can probably be improved:

| Action | Comment | Message |
|---|---|---|
| Add | | _User added 7TV emote AlienPls. [emote preview]_ |
| Add | disabled unlisted emtoes | _User added 7TV emote AlienPls, but you disabled unlisted emotes._ |
| Update | renamed | _User updated 7TV emote AlienPlsAlias (AlienPls)._ |
| Update | renamed (not added to channel/unlisted emote) | _User updated 7TV emote AlienPlsAlias (AlienPls), but it's not added._ |
| Update | other | _User updated 7TV emote AlienPls._ |
| Update | other (not added to channel/unlisted emote) | _User updated 7TV emote AlienPls, but it's not added._ |
| Remove | | _User removed 7TV emote AlienPls._ |
| Remove | not added clientside (most likely unlisted) | _User removed 7TV emote AlienPls, but it wasn't added._ |

AnatoleAM commented:
> If the eventapi will stay the same

I'll confirm that `/v1` won't change; it will send the same data when modifying the connected emote set (the v3 equivalent of channel emotes). Though note that due to v1 events only supporting single emote changes, this means that when an entire emote set is swapped on a channel, it will send hundreds of events which may heavily spam the chat. So it might be worth adding an option to disable the messages if this bothers some users.

If you'd like to track the ongoing implementation of EventAPI v3 (yes, v2 got skipped so that all apis have the same version) there is an open PR: https://github.com/SevenTV/EventAPI/pull/58

TroyKomodo committed:
add settings

TroyKomodo committed:
remove websocket reference

TroyKomodo committed:
add buffer system

TroyKomodo committed:
fix format + minor nits

TroyKomodo committed:
fix and spacing

Nerixyz committed:
fix: rename `MAX_PING_SECONDS` to the more meaningful `CHECK_PING_INTERVAL`

TroyKomodo commented on src/Application.cpp:
```diff
@@ -151,6 +151,12 @@ void Application::initialize(Settings &settings, Paths &paths)
+
+    if (settings.enableSevenTVEventApi)
+    {
+        qWarning("enabled event api");
```
Can we remove this

Nerixyz replied:
Yes

Nerixyz committed:
fix: remove `qWarning` log

This is redundant, because the manager will log when it starts.

williambulin commented on src/common/Channel.cpp:
```diff
@@ -191,6 +191,110 @@ void Channel::addOrReplaceTimeout(MessagePtr message)
+                     MessageFlag::SevenTvEventApiRemoveEmoteMessage,
+                     MessageFlag::SevenTvEventApiUpdateEmoteMessage}))
+            {
+                auto found = false;
```
You could use the stl here:
```cpp
#include <algorithm>

bool found{std::any_of(vec1.begin(), vec1.end(), [&vec2](auto const &x) {
  return std::find(vec2.begin(), vec2.end(), x) != vec2.end();
}))};
```

williambulin commented on src/providers/twitch/TwitchChannel.cpp:
```diff
@@ -1246,4 +1298,12 @@ boost::optional<CheerEmote> TwitchChannel::cheerEmote(const QString &string)
 
+void TwitchChannel::listenSeventv()
+{
+    if (getApp()->twitch->eventApi)
```
If you wanna be consistent with other changes you've made, you could do this to avoid another call to getApp and avoid repetition
```cpp
if (auto eventApi = getApp()->twitch->eventApi)
        eventApi->joinChannel(this->getName());
```
but that's honestly just nitpicking

williambulin commented on src/common/Channel.cpp:
```diff
@@ -191,6 +191,110 @@ void Channel::addOrReplaceTimeout(MessagePtr message)
     // WindowManager::instance().repaintVisibleChatWidgets(this);
 }
 
+void Channel::addOrReplaceSevenTvEventAddRemove(MessagePtr message)
```
I'm not aware of the context of the PR but this function name is honestly pretty hard to understand

Nerixyz replied:
We're following the `addOrReplaceTimeout` method name.

williambulin commented on src/providers/seventv/SeventvEmotes.cpp:
```diff
@@ -139,6 +145,57 @@ namespace {
+        //         This is expensive but the cache entry may be corrupted
+        //         when an emote was added with a different alias in some other
+        //         channel.
+        for (auto it = map->begin(); it != map->end(); it++)
```
Why not use a ranged for here? You can just dereference the shared_ptr to get a reference to the EmoteMap and then use it accordingly in the entire function

Nerixyz committed:
fix: use `std::any_of(.., std::find)` to check messages

williambulin commented on src/providers/twitch/TwitchIrcServer.cpp:
```diff
@@ -37,6 +39,15 @@ TwitchIrcServer::TwitchIrcServer()
 
+    if (getSettings()->enableSevenTVEventApi)
+    {
+        this->eventApi = new SeventvEventApi(SEVENTV_EVENTAPI_URL);
```
Where is the
```cpp
delete
```
to free the memory, why not just use a smartptr here?

TroyKomodo replied:
This is never deleted

TroyKomodo replied:
It is never deallocated because once running we don't stop it, the setting to disable it requires a restart.

Nerixyz committed:
fix: simplify always true condition

williambulin commented on src/providers/twitch/TwitchIrcServer.cpp:
```diff
@@ -37,6 +39,15 @@ TwitchIrcServer::TwitchIrcServer()
+    }
+    else
+    {
+        this->eventApi = nullptr;
```
The whole condition is not needed, just initialize the variable in the class/header to nullptr

williambulin commented on src/providers/twitch/TwitchIrcServer.hpp:
```diff
@@ -42,6 +44,7 @@ class TwitchIrcServer final : public AbstractIrcServer, public Singleton
     IndirectChannel watchingChannel;
 
     PubSub *pubsub;
+    SeventvEventApi *eventApi;
```
Initialize to nullptr here directly (related to the comment in the related .cpp file) and/or swap this with a smart pointer anyway.

Nerixyz replied:
> and/or swap this with a smart pointer anyway.

I agree. I've been following `PubSub`, which uses a c-style pointer.

Changed it to `unique_ptr`.

williambulin commented on src/providers/twitch/TwitchChannel.cpp:
```diff
@@ -572,6 +580,50 @@ std::shared_ptr<const EmoteMap> TwitchChannel::ffzEmotes() const
+{
+    auto removed =
+        SeventvEmotes::removeEmote(this->seventvEmotes_, action.emoteName);
+    if (removed)
```
```cpp
    if (auto removed =
        SeventvEmotes::removeEmote(this->seventvEmotes_, action.emoteName))
```
For consistency with the rest of the code

williambulin replied:
Note: you don't even use removed anyway so might as well not have it
```cpp
if (SeventvEmotes::removeEmote(this->seventvEmotes_, action.emoteName))
```

williambulin commented on src/providers/twitch/TwitchChannel.cpp:
```diff
@@ -572,6 +580,50 @@ std::shared_ptr<const EmoteMap> TwitchChannel::ffzEmotes() const
+    auto baseName = action.emote->baseName;
+    auto result = SeventvEmotes::updateEmote(this->seventvEmotes_, &baseName,
+                                             action.emote->json);
+    if (result.has_value())
```
Code flow will be better if you invert that condition and return
```cpp
if (!result.has_value())
  return;
```

williambulin commented on src/providers/twitch/TwitchChannel.cpp:
```diff
@@ -185,6 +185,13 @@ TwitchChannel::TwitchChannel(const QString &name)
     });
 
+    this->destroyed.connect([this]() {
+        if (getApp()->twitch->eventApi)
```
Same comment as in another file, for consistency:
```cpp
if (auto eventApi = getApp()->twitch->eventApi)
        eventApi->joinChannel(this->getName());
```

Nerixyz committed:
fix: don't call `getApp` twice

Nerixyz committed:
fix: use range-base for-loop

Nerixyz committed:
fix: avoid calling `getApp` twice

Nerixyz committed:
fix: use early-return when updating fails

Nerixyz committed:
fix: use early-return when removing fails

Nerixyz committed:
refactor: use `unique_ptr` for `SeventvEventApi`

TroyKomodo commented on src/providers/twitch/TwitchChannel.cpp:
```diff
@@ -616,13 +616,14 @@ void TwitchChannel::updateSeventvEmote(const EventApiEmoteUpdate &action)
-            MessageBuilder(seventvRemoveEmoteMessage, action.actor,
-                           {action.emoteName})
-                .release());
+    if (!SeventvEmotes::removeEmote(this->seventvEmotes_, action.emoteName))
```
Invert this early return

Nerixyz committed:
fix: invert early return

AnatoleAM commented:
@Nerixyz Thanks for your contribution to our branch of the Chatterino project. I assigned you the Contributor chat badge, which you can turn it on within the Wardrobe on 7tv.app (top right, golden hanger icon).
